test/util/IdTableHelpers.cpp

```diff
@@ -0,0 +1,217 @@
+
+#include "engine/idTable/IdTable.h"
+#include "global/ValueId.h"
+#include "util/Algorithm.h"
+#include "util/Exception.h"
+
+// ____________________________________________________________________________
+void compareIdTableWithExpectedContent(
+    const IdTable& table, const IdTable& expectedContent,
+    const bool resultMustBeSortedByJoinColumn, const size_t joinColumn,
+    ad_utility::source_location l) {
+  // For generating more informative messages, when failing the comparison.
+  std::stringstream traceMessage{};
+
+  auto writeIdTableToStream = [&traceMessage](const IdTable& idTable) {
+    std::ranges::for_each(idTable,
+                          [&traceMessage](const auto& row) {
+                            // TODO<C++23> Use std::views::join_with for both
+                            // loops.
+                            for (size_t i = 0; i < row.numColumns(); i++) {
+                              traceMessage << row[i] << " ";
+                            }
+                            traceMessage << "\n";
+                          },
+                          {});
+  };
+
+  traceMessage << "compareIdTableWithExpectedContent comparing IdTable\n";
+  writeIdTableToStream(table);
+  traceMessage << "with IdTable \n";
+  writeIdTableToStream(expectedContent);
+  auto trace{generateLocationTrace(l, traceMessage.str())};
+
+  // Because we compare tables later by sorting them, so that every table has
+  // one definit form, we need to create local copies.
+  IdTable localTable{table.clone()};
+  IdTable localExpectedContent{expectedContent.clone()};
+
+  if (resultMustBeSortedByJoinColumn) {
+    // Is the table sorted by join column?
+    ASSERT_TRUE(std::ranges::is_sorted(localTable.getColumn(joinColumn)));
+  }
+
+  // Sort both the table and the expectedContent, so that both have a definite
+  // form for comparison.
+  std::ranges::sort(localTable, std::ranges::lexicographical_compare);
+  std::ranges::sort(localExpectedContent, std::ranges::lexicographical_compare);
+
+  ASSERT_EQ(localTable, localExpectedContent);
+}
+
+// ____________________________________________________________________________
+void sortIdTableByJoinColumnInPlace(IdTableAndJoinColumn& table) {
+  CALL_FIXED_SIZE((std::array{table.idTable.numColumns()}), &Engine::sort,
+                  &table.idTable, table.joinColumn);
+}
+
+// ____________________________________________________________________________
+IdTable generateIdTable(
+    const size_t numberRows, const size_t numberColumns,
+    const std::function<std::vector<ValueId>()>& rowGenerator) {
+  AD_CONTRACT_CHECK(numberRows > 0 && numberColumns > 0);
+
+  // Creating the table and setting it to the wanted size.
+  IdTable table{numberColumns, ad_utility::testing::makeAllocator()};
+  table.resize(numberRows);
+
+  // Fill the table.
+  for (auto restrictedRow = table.begin(); restrictedRow != table.end();
+       restrictedRow++) {
+    // Make sure, that the generated row has the right
```

Comment:
Those are not `restrictedRow`s but `iterators`. So either call those `it` or `rowIt` or something like this.

Reply:
No longer relevant.